ENH: Adding in proportion plots

mortonjt opened the pull request:
This also changes the API for balance taxonomy, since it now requires a composition
table as input instead of balances as input.

This also adds in a parameter `threshold` that discretizes continuous metadata to
plot in the proportion plot.

This is an example of what it looks like

![image](https://user-images.githubusercontent.com/4184797/31695752-55492548-b361-11e7-98c6-638bb819be9f.png)

And we updated the colors for the taxonomy barplot to make sure that numerator and denominator can be differentiated.

mortonjt committed:
ENH: Adding in proportion plots

This also changes the API for balance taxonomy, since it now requires a composition
table as input instead of balances as input.

This also adds in a parameter `threshold` that discretizes continuous metadata to
plot in the proportion plot

mortonjt committed:
STY: pep8 / flake8

mortonjt commented:
Looks like we need an actual conda release for the newest version of gneiss

ebolyen commented:
@mortonjt, what would be a timeline for a release on biocore?

mortonjt commented:
As soon as @RNAer double checks the PR [here](https://github.com/biocore/conda-recipes/pull/107).  The build is actually working -- but it may take a few hours to actually upload it once it is ready.

mortonjt commented:
Alternatively, we could rely on the pypi install -- its already on pypi now.

mortonjt commented:
Ok, the conda recipe is in!

mortonjt committed:
Merge branch 'master' of https://github.com/qiime2/q2-gneiss into proportion-plots

mortonjt committed:
Adding warning messages

jairideout commented:
@mortonjt coverage dropped 3.3%, can you look into that please?

mortonjt committed:
TST: Adding in tests for warnings

mortonjt committed:
TST: Adding in relevant test datasets

mortonjt commented:
@jairideout thank you for catching that.  The tests + data files have been added to cover the new warning functionality.

ebolyen commented:
looks like there's a lint error regarding warnings import (now that you raise instead)

mortonjt committed:
flake8

mortonjt committed:
TST: Adding test for multiple categories

mortonjt commented:
@ebolyen what do you think now?

ebolyen commented on q2_gneiss/plot/_plot.py:
```diff
@@ -110,6 +188,13 @@ def f(x):
+        if dcat is not None:
+            index_f.write('<h1>Proportion Plot </h1>\n')
+            index_f.write(('<img src="proportion_plot.svg" '
+                           'alt="proportionss">\n\n'
```
typo `proportions`

mortonjt replied:
Fixed!

ebolyen commented on q2_gneiss/plot/_plot.py:
```diff
@@ -49,56 +55,128 @@ def f(x):
+                else '%s > %f' % (c.name, threshold)
+            )
+
         except:
```
Not strictly related, but since you are here, could you change this to `except Exception:`?

mortonjt replied:
Done!

ebolyen commented on q2_gneiss/plot/_plot.py:
```diff
@@ -49,56 +55,128 @@ def f(x):
+                     'Proportion plots will not be displayed'),
+                    stacklevel=2)
+            else:
+                dcat = c
```
Could we check to see if `threshold` is not `None` when we are in this branch? As the threshold isn't really doing anything, but the user may still have mistakenly passed it (or they are using the wrong column).

mortonjt replied:
Done!

ebolyen commented on q2_gneiss/plot/_plot.py:
```diff
@@ -49,56 +55,128 @@ def f(x):
+
+    balances = balances * np.sqrt(r * s / (r + s))
+    balances = pd.DataFrame(balances, index=table.index,
+                            columns=[balance_name])
```
I don't completely understand this stuff still, but the sqrt looks like the IRL transform? Is there some code we could reuse here instead? I guess I don't really understand why the input went for `Balance` to `Composition`, but I take it something special is happening here?

ebolyen replied:
OH, it's using the taxonomy as the basis, which is the new part here, right?

mortonjt replied:
Not quite - this PR is completely independent from #18

The main motivation here is to reduce the number for free floating parameters.

We do need the `--i-table` here for the proportion plots.  Rather than having the user pass in all of the balances as an extra parameter, I was thinking that it may be easier to just compute the ilr transform that one balance that the user is interested in (since it is super fast).

ebolyen commented on q2_gneiss/plot/_plot.py:
```diff
@@ -130,20 +215,27 @@ def f(x):
+        'metadata': 'Metadata for plotting the balance (optional).',
+        'n_features': 'The number of features to plot in the proportion plot.',
+        'threshold': ('A threshold to designate discrete categories '
+                      'for continuous metadata.  If not specified, '
```
This was pretty confusing until I realized it was splitting your metadata into two groups. Maybe mentioning that would be good, and also whether the behavior is inclusive or exclusive with respect to the threshold and the first group.

mortonjt replied:
Updating the `threshold` parameter as 
```
        'threshold': ('A threshold to designate discrete categories '
                      'for numerical metadata.  This will split the '
                      'numerical category into two categories, values below '
                      'the threshold, and values above the threshold.'
                      'If not specified, this threshold will '
                      'default to the mean.')
```

ebolyen reviewed (changes requested):
Minor comments.

ebolyen commented:
Also, I'm probably wrong about this, but I think the colors may be misleading. In the proportion plot, there's basically 4 quadrants, numerator-group1, numerator-group2, denominator-group1, and denominator-group2. But in the balance taxonomy, the barplot colors match the group colors instead of the numerator/denominator colors.

mortonjt commented:
@ebolyen you raised a very valid point.  I honestly wasn't sure about the colors.  The numerator / denominator colors are coloring features, whereas the samples have a separate coloring scheme.

This is definitely something worthwhile to think about in the context of qiime2 as a whole, and how we want to propagate these values as DataDictionaries.  Or how we should link sample colors to the feature colors.  At the moment, I don't know how to do this.  But maybe we can open these up as input parameters in the mean time (i.e. specify `num_color`, `denom_color`, and a separate parameter specifying the seaborn color wheel for the samples).

ebolyen commented:
> @ebolyen you raised a very valid point. I honestly wasn't sure about the colors. The numerator / denominator colors are coloring features, whereas the samples have a separate coloring scheme.

Sorry, I like the split in colors between the sample groups and the features, but I think the numerator/denominator coloring scheme is actually used as group colors in the new proportion plot (left vs right, not top vs bottom), but it *is* used as the numerator/denominator colors in the balance taxonomy plot.

> This is definitely something worthwhile to think about in the context of qiime2 as a whole, and how we want to propagate these values as DataDictionaries. Or how we should link sample colors to the feature colors. At the moment, I don't know how to do this. But maybe we can open these up as input parameters in the mean time (i.e. specify num_color, denom_color, and a separate parameter specifying the seaborn color wheel for the samples).

Agreed! Although I don't think parameterizing it is necessary at this point.

ebolyen commented:
oh wait, I think I understand now, in the proportion plot, we're comparing a proportion of balances. So while the balances *have* a numerator and denominator, that isn't the same one we are looking at here. The balance's numerator and denominator are the shaded areas in the plot right?

EDIT: nevermind, the axis are independent, I don't really understand what the plot is saying now.

mortonjt committed:
FIX: Fixing @ebolyen comments.

Also adjusting color schemes to make samples and features distinct.
Ultimately this required another way to sort the features.

mortonjt commented:
:( 
Below are images with the new color schemes we discussed.  What do you think?
![image](https://user-images.githubusercontent.com/4184797/31923436-519eb30e-b82f-11e7-9403-cb177d678400.png)

![image](https://user-images.githubusercontent.com/4184797/31923440-5a0340c8-b82f-11e7-9dac-eac8e975fd9f.png)


![image](https://user-images.githubusercontent.com/4184797/31923438-54ce403a-b82f-11e7-89d9-821ebab8e509.png)

ebolyen commented:
@mortonjt, why the sad face? I think these colors are much more interpretable! 👍

mortonjt committed:
ENH: pep8

mortonjt committed:
FIX: fixing issue with ordering numerator / denominator

mortonjt committed:
FIX: fixing continuous usecase

mortonjt committed:
removing prints

mortonjt committed:
Fixing typo in top_denom_features

mortonjt committed:
TST: Adding check for threshold and adding another test

mortonjt committed:
pep8

mortonjt committed:
DOC: Updating CLI docstring for threshold

mortonjt commented:
@ebolyen thank you for the thorough review.  This was extremely helpful.  I believe that the major comments have been addressed.  Thanks!!!

ebolyen commented on q2_gneiss/plot/_plot.py:
```diff
@@ -122,6 +122,11 @@ def f(x):
+            if threshold is not None:
+                raise ValueError('Categorical data detected. '
+                                 'Only specify a threshold for '
+                                 'numerical valued metadata')
```
`numerical valued metadata` -> `numeric metadata.`

ebolyen commented on q2_gneiss/plot/_plot.py:
```diff
@@ -173,7 +178,8 @@ def f(x):
             metadata = pd.DataFrame({dcat.name: dcat})
             top_num_features = num_fold_change.index[-n_features:]
-            top_denom_features = num_fold_change.index[:n_features]
+            top_denom_features = denom_fold_change.index[:n_features]
```
I think this is still taking the bottom of the sort, should it be `[-n_features:]`?

ebolyen commented on q2_gneiss/plot/tests/test_plot.py:
```diff
@@ -295,6 +295,13 @@ def test_balance_taxonomy_categorical(self):
+            balance_taxonomy(self.results, self.table, self.tree,
+                             self.taxonomy, balance_name='a',
+                             metadata=self.categorical,
+                             threshold=100.)
```
👍

mortonjt committed:
FIX: fixing another typo wrt to numerator and denominator
